pkg/webhook/server/certificate/manager.go

```diff
@@ -55,123 +54,152 @@ const (
 // It will also update the webhook caBundle field with the cluster CA cert and
 // approve the generated cert/key with k8s certification approval mechanism
 func NewManager(
-	crMgr crmanager.Manager,
+	client client.Client,
 	webhookName string,
 	webhookType WebhookType,
-	certDir string, certFileName string, keyFileName string, caConfigMapKey types.NamespacedName, caConfigMapField string) (*manager, error) {
-
-	certStore, err := NewFilePairStore(
-		certDir,
-		certFileName,
-		keyFileName,
-	)
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to initialize new webhook cert/key file store")
-	}
+) *manager {
 
 	m := &manager{
-		crMgr:            crMgr,
-		log:              logf.Log.WithName("webhook/server/certificate/manager"),
-		certStore:        certStore,
-		caConfigMapKey:   caConfigMapKey,
-		caConfigMapField: caConfigMapField,
+		stopCh:           make(chan struct{}),
+		client:           client,
+		webhookName:      webhookName,
+		webhookType:      webhookType,
+		keyPairByService: map[types.NamespacedName]*triple.KeyPair{},
+		now:              time.Now,
+		ready:            false,
+		log: logf.Log.WithName("webhook/server/certificate/manager").
+			WithValues("webhookType", webhookType, "webhookName", webhookName),
 	}
+	return m
+}
 
-	dnsNames := []string{}
-	if webhookType == MutatingWebhook {
-		mutatingWebHookConfig, err := m.updateMutatingWebhookCABundle(webhookName)
-		if err != nil {
-			return nil, errors.Wrap(err, "failed to update CA bundle at webhook")
+// Will start the the underlaying client-go cert manager [1]  and
+// wait for TLS key and cert to be generated
+//
+// [1] https://godoc.org/k8s.io/client-go/util/certificate
+func (m *manager) Start() error {
+	m.log.Info("Starting cert manager")
+
+	go wait.Until(func() {
+		deadline := m.nextRotationDeadline()
+		if sleepInterval := deadline.Sub(m.now()); sleepInterval > 0 {
+			m.log.Info(fmt.Sprintf("Waiting %v for next certificate rotation", sleepInterval))
+
+			timer := time.NewTimer(sleepInterval)
+			defer timer.Stop()
+
+			select {
+			case <-timer.C:
+			}
 		}
 
-		for _, webhook := range mutatingWebHookConfig.Webhooks {
-			dnsNames = append(dnsNames, dnsNamesForService(*webhook.ClientConfig.Service)...)
+		backoff := wait.Backoff{
+			Duration: 2 * time.Second,
+			Factor:   2,
+			Jitter:   0.1,
+			Steps:    5,
 		}
-	} else if webhookType == ValidatingWebhook {
-		validatingWebHookConfig, err := m.updateValidatingWebhookCABundle(webhookName)
-		if err != nil {
-			return nil, errors.Wrap(err, "failed to update CA bundle at webhook")
+		if err := wait.ExponentialBackoff(backoff, m.rotateCondition); err != nil {
+			utilruntime.HandleError(fmt.Errorf("Reached backoff limit, still unable to rotate certs: %v", err))
+			wait.PollInfinite(32*time.Second, m.rotateCondition)
```

Comment:
If we make the pod restart by panic or by returning error at `Start` [1] it will crash whatever the pod is running, if we do so we have to document that cert-manager has to be run at separate `Deployment`. Is safer to just use HandlerError also I can simplify this since this is from the old CSR implementation that involves cluster communication here we just call some crypto primitives, I think PollImmediate is enough.

[1] https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/manager/manager.go#L225